common/logger.go

```diff
@@ -84,23 +92,69 @@ func (l *Logger) Logf(level logrus.Level, category string, msg string, args ...i
 	if now == elapsed {
 		elapsed = 0
 	}
+	defer func() {
+		l.lastLogCall = now
+	}()
 
-	if l.categoryFilter == nil || l.categoryFilter.Match([]byte(category)) {
+	if l.categoryFilter != nil && !l.categoryFilter.Match([]byte(category)) {
+		return
+	}
+	if l.log == nil {
 		magenta := color.New(color.FgMagenta).SprintFunc()
-		if l.logger != nil {
-			entry := l.logger.WithFields(logrus.Fields{
-				"category":  magenta(category),
-				"elapsed":   fmt.Sprintf("%d ms", elapsed),
-				"goroutine": goRoutineID(),
-			})
-			if l.logger.GetLevel() < level && l.debugOverride {
-				entry.Printf(msg, args...)
-			} else {
-				entry.Logf(level, msg, args...)
-			}
-		} else {
-			fmt.Printf("%s [%d]: %s - %s ms\n", magenta(category), goRoutineID(), string(msg), magenta(elapsed))
+		fmt.Printf("%s [%d]: %s - %s ms\n", magenta(category), goRoutineID(), string(msg), magenta(elapsed))
+		return
+	}
+	entry := l.log.WithFields(logrus.Fields{
+		"category":  category,
+		"elapsed":   fmt.Sprintf("%d ms", elapsed),
+		"goroutine": goRoutineID(),
+	})
+	if l.log.GetLevel() < level && l.debugOverride {
+		entry.Printf(msg, args...)
+		return
+	}
+	entry.Logf(level, msg, args...)
+}
+
+// SetLevel sets the logger level from a level string.
+// Accepted values:
+func (l *Logger) SetLevel(level string) error {
+	pl, err := logrus.ParseLevel(level)
+	if err != nil {
+		return err
+	}
+	l.log.SetLevel(pl)
+	return nil
+}
+
+// DebugMode returns true if the logger level is set to Debug or higher.
+func (l *Logger) DebugMode() bool {
+	return l.log.GetLevel() >= logrus.DebugLevel
+}
+
+// ReportCaller adds source file and function names to the log entries.
+func (l *Logger) ReportCaller() {
+	caller := func() func(*runtime.Frame) (string, string) {
+		return func(f *runtime.Frame) (function string, file string) {
+			return f.Func.Name(), fmt.Sprintf("%s:%d", f.File, f.Line)
 		}
-		l.lastLogCall = now
 	}
+	l.log.SetFormatter(&logrus.TextFormatter{
+		CallerPrettyfier: caller(),
+		FieldMap: logrus.FieldMap{
+			logrus.FieldKeyFile: "caller",
+		},
+	})
+	l.log.SetReportCaller(true)
```

Comment:
Do we need this custom field and prettifier? From what I see it only changes the field name.

With:
```
func="github.com/grafana/xk6-browser/common.(*Logger).Logf" caller="github.com/grafana/xk6-browser@v0.0.0-00010101000000-000000000000/common/logger.go:116"
```

Without (just calling `l.log.SetReportCalled(true)`):
```
func="github.com/grafana/xk6-browser/common.(*Logger).Logf" file="github.com/grafana/xk6-browser@v0.0.0-00010101000000-000000000000/common/logger.go:116"
```

Is the idea to leave it for further customization (cleaning up the `github.com/...` prefix like you mentioned)?

Also, there are a lot of places where the caller is `Logger.Logf`. That might be tricky to resolve, so this is fine as is.

Reply:
Yeah, it's because we're using our custom logger. Maybe we should get rid of it as you said before.

Reply:
Yeah, it's because we're using our custom logger. Maybe we should get rid of it as you said before.

For now, I can do something like this, WDYT?

```golang
// ReportCaller adds source file and function names to the log entries.
func (l *Logger) ReportCaller() {
	// strip the module informaiton
	strip := func(s string) string {
		s = strings.TrimPrefix(s, "github.com/grafana/xk6-browser")
		s = s[strings.Index(s, "/")+1:]

		return s
	}
	caller := func() func(*runtime.Frame) (string, string) {
		return func(f *runtime.Frame) (fn string, loc string) {
			loc = strip(fmt.Sprintf("%s:%d", f.File, f.Line))
			// find the caller of the log func
			fn = f.Func.Name()

			const stackLevel = 8
			_, file, no, ok := runtime.Caller(stackLevel)
			if ok {
				fn = fmt.Sprintf("%s:%d", strip(file), no)
			}

			return fn, loc
		}
	}
	l.log.SetFormatter(&logrus.TextFormatter{
		CallerPrettyfier: caller(),
		FieldMap: logrus.FieldMap{
			logrus.FieldKeyFile: "caller",
		},
	})
	l.log.SetReportCaller(true)
}
```

Output:
```
time="2021-11-30T12:12:21+03:00" level=debug msg="sid:AFA9A388DA1617DF05688B6718BCD591 tid:045F8A2BB4D7E102A47820CF00E5C9DF" func="common/session.go:75" caller="common/logger.go:116" category="Session:close" elapsed="0 ms" goroutine=68
```

I can easily strip out the `caller` there as well.

@imiric

Reply:
Yeah, we wouldn't need `XK6_BROWSER_CALLER` if the caller information could be added by default on levels >= `Debug`.

Can that `stackLevel = 8` be a constant? I mean, will that always be the case? I don't have time to look into it now, but if you think it solves it for all cases, LGTM.

And the `func="common/session.go:75"` doesn't make much sense, there's no function name :) Maybe this should be `caller="common.(*Session).close [common/session.go:75]"`?

In any case, don't spend too much time on this. Feel free to create an issue for it and we can tackle it soon-ish.

Reply:
Yep, we need function name there as well, but at least it goes to the file and line when I click on it on my terminal 😅 

I created another issue with some rough PRs that solve some of the issues:
* Use logger directly: #137 
* PR: #138

@imiric

Reply:
> Can that stackLevel = 8 be a constant? I mean, will that always be the case? I don't have time to look into it now, but if you think it solves it for all cases, LGTM.

For this Logrus version, yes!